### PR TITLE
Refactor XMPP module , fixes #254

### LINKED ISSUE #254: Refactor XMPP module
Plinth can potentially support multiple XMPP servers and multiple XMPP web clients. To encourage this, I suggest the following changes:
1. Split xmpp module into 2 modules: Chat Server (Ejabberd) and Chat Client (JWChat).
2. Name / URL of each new module should match the package's name.

### PR DESCRIPTION
**Pull request by 500swapnil**

### CODE REVIEW

**Comment by 500swapnil** [comment]
I have made separate modules for jsxc and ejabberd using the xmpp code. How do I isolate jsxc from ejabberd and vice-versa?

**Comment by SunilMohanAdapa** [comment]
* Get rid of static/ directory from ejabberd module.
* Rename ejabberd/templates/xmpp.html to ejabberd/templates/ejabberd.html
* Rename jsxc/templates/jsxc.html to jsxc/templates/jsxc_launch.html
* Rename jsxc/templates/xmpp.html to jsxc/templates/jsxc.html
* Remove jsxc URL from ejabberd module.
* Remove xmpp URL from JSXC module.
* Remove JSXC view from ejabberd module.
* Rename EjabberedServiceView to JSXCServiceView in jsxc module.
* Similarly split managed_packages, description, shortcuts code.
* Rename the xmpp action to ejabberd.

Finally, there should no hint of JSXC in the ejabberd module.  There should be no hint of ejabberd/xmpp in jsxc module.  Everything that was called xmpp should now be called ejabberd.

ejabberd module will install and configure ejabberd.  jsxc module will install and launch jsxc.

**Comment by 500swapnil** [comment]
Separated the modules. Let me know if any changes are required.

File: plinth/modules/jsxc/__init__.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+from plinth import service as service_module
+from plinth.signals import pre_hostname_change, post_hostname_change
+from plinth.signals import domainname_change
+
+
+version = 2
+
+depends = ['apps']
+
+managed_services = ['ejabberd']
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: jsxc module should not manage ejabberd service.

File: plinth/modules/jsxc/__init__.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+"""
+Plinth module to configure XMPP server
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Please change XMPP server -> XMPP web client / jsxc.

File: plinth/modules/jsxc/__init__.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+from plinth.signals import domainname_change
+
+
+version = 2
+
+depends = ['apps']
+
+managed_services = ['ejabberd']
+
+managed_packages = ['libjs-jsxc', 'ejabberd']
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: JSXC module shouldn't manage ejabberd package, I guess.

File: plinth/modules/jsxc/__init__.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+title = _('Chat Server \n (XMPP)')
+
+description = [
+    _('XMPP is an open and standardized communication protocol. Here '
+      'you can run and configure your XMPP server, called ejabberd.'),
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Please remove ejabberd stuff.

File: plinth/modules/jsxc/__init__.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+def init():
+    """Initialize the XMPP module"""
+    menu = cfg.main_menu.get('apps:index')
+    menu.add_urlname(title, 'glyphicon-comment', 'xmpp:index')
+
+    global service
+    setup_helper = globals()['setup_helper']
+    if setup_helper.get_state() != 'needs-setup':
+        service = service_module.Service(
+            'ejabberd', title,
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: jsxc will have its own service mapped to http/https.

File: plinth/modules/jsxc/__init__.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            disable=disable)
+    helper.call('post', service.notify_enabled, None, True)
+    helper.call('post', add_shortcut)
+
+
+def add_shortcut():
+    frontpage.add_shortcut('jsxc', _('Chat Client \n (jsxc)'),
+                           url=reverse_lazy('xmpp:jsxc'),
+                           login_required=True)
+    frontpage.add_shortcut('xmpp', title,
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: remove xmpp shortcut

File: plinth/modules/jsxc/__init__.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                          ['change-domainname',
+                           '--domainname', new_domainname],
+                          async=True)
+
+
+def diagnose():
+    """Run diagnostics and return the results."""
+    results = []
+
+    results.append(action_utils.diagnose_port_listening(5222, 'tcp4'))
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: All of these diagnostics are for ejabberd, not jsxc.

File: plinth/modules/jsxc/templates/jsxc.html [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+{% load i18n %}
+
+{% block description %}
+
+  {% for paragraph in description %}
+    <p>{{ paragraph|safe }}</p>
+  {% endfor %}
+
+  <p>
```
Comment (jvalleroy) [inline plinth/modules/jsxc/templates/jsxc.html]: remove this paragraph

File: plinth/modules/jsxc/__init__.py (hunk 5 above)
Comment (500swapnil) [inline plinth/modules/jsxc/__init__.py]: Would replacing the service_module.Service(..) with http work?

Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: @500swapnil It should be similar to other web apps, for example:
https://github.com/freedombox/Plinth/blob/master/plinth/modules/roundcube/__init__.py#L72

Comment (500swapnil) [inline plinth/modules/jsxc/__init__.py]: Okay and the title being 'jsxc'

**Comment by 500swapnil** [comment]
@jvalleroy I have made the changes.

File: plinth/modules/jsxc/__init__.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+from plinth.signals import domainname_change
+
+
+version = 2
+
+depends = ['apps']
+
+managed_packages = ['libjs-jsxc']
+
+title = _('Chat Server \n (XMPP)')
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Need to change title.

File: plinth/modules/jsxc/__init__.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+service = None
+
+logger = logging.getLogger(__name__)
+
+
+def init():
+    """Initialize the XMPP module"""
+    menu = cfg.main_menu.get('apps:index')
+    menu.add_urlname(title, 'glyphicon-comment', 'xmpp:index')
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Need to update menu url.

File: plinth/modules/jsxc/__init__.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+    global service
+    setup_helper = globals()['setup_helper']
+    if setup_helper.get_state() != 'needs-setup':
+        service = service = service_module.Service(
+            'jsxc', title, ports=['http', 'https'], is_external=True,
+            is_enabled=is_enabled, enable=enable, disable=disable)
+        if is_enabled():
+            add_shortcut()
+    pre_hostname_change.connect(on_pre_hostname_change)
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Ejabberd will handle hostname / domain changes, don't need to repeat it here.

File: plinth/modules/jsxc/__init__.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+def setup(helper, old_version=None):
+    """Install and configure the module."""
+    domainname = get_domainname()
+    logger.info('XMPP service domainname - %s', domainname)
+
+    helper.call('pre', actions.superuser_run, 'xmpp',
+                ['pre-install', '--domainname', domainname])
+    helper.install(managed_packages)
+    helper.call('post', actions.superuser_run, 'xmpp', ['setup'])
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Need to separate jsxc setup from ejabberd setup. Will probably require changing or splitting actions/xmpp.

File: plinth/modules/jsxc/__init__.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    global service
+    if service is None:
+        service = service_module.Service(
+            'ejabberd', title,
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Should be jsxc service.

File: plinth/modules/jsxc/__init__.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+def add_shortcut():
+    frontpage.add_shortcut('jsxc', _('Chat Client \n (jsxc)'),
+                           url=reverse_lazy('xmpp:jsxc'),
+                           login_required=True)
+
+
+def is_enabled():
+    """Return whether the module is enabled."""
+    return (action_utils.service_is_enabled('ejabberd') and
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Should only check if jsxc web conf is enabled. (Remove jwchat.)

File: plinth/modules/jsxc/__init__.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+
+def get_domainname():
+    """Return the domainname"""
+    fqdn = socket.getfqdn()
+    return '.'.join(fqdn.split('.')[1:])
+
+
+def enable():
+    """Enable the module."""
+    actions.superuser_run('xmpp', ['enable'])
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Should enable jsxc web conf only.

File: plinth/modules/jsxc/__init__.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    add_shortcut()
+
+
+def disable():
+    """Enable the module."""
+    actions.superuser_run('xmpp', ['disable'])
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Should disable jsxc web conf only.

File: plinth/modules/jsxc/__init__.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    frontpage.remove_shortcut('jsxc')
+    frontpage.remove_shortcut('xmpp')
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Should not remove xmpp shortcut.

File: plinth/modules/jsxc/__init__.py (hunk 10 above)
Comment (500swapnil) [inline plinth/modules/jsxc/__init__.py]: What should be the new menu url?

File: plinth/modules/jsxc/__init__.py (hunk 12 above)
Comment (500swapnil) [inline plinth/modules/jsxc/__init__.py]: How do I separate these?

**Comment by 500swapnil** [comment]
@jvalleroy I have made all the changes other than splitting the setup and menu url.

File: plinth/modules/jsxc/__init__.py (hunk 10 above)
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: @500swapnil For ejabberd module, it should be 'ejabberd:index'.
For jsxc module, 'jsxc:index'.

File: plinth/modules/jsxc/__init__.py (hunk 12 above)
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: @500swapnil Actually, we can just rename actions/xmpp to actions/ejabberd.
We don't need to run this setup when jsxc module is installed though.

**Comment by 500swapnil** [comment]
@jvalleroy Done!

**Comment by SunilMohanAdapa** [comment]
On Thursday 05 January 2017 07:33 PM, James Valleroy wrote:
> *@jvalleroy* commented on this pull request.
> 
> I think Sunil will review this later, but for now I pointed out some
> places where ejabberd is still mixed up with jsxc module.
> 

@jvalleroy , Please feel free to comment, review and commit this work or
another works that I am involved in.  My schedule will still take at
least a couple of weeks to become normal.

-- 
Sunil

File: plinth/modules/ejabberd/urls.py [hunk 18]
```diff
@@ -16,15 +16,14 @@
...
 from django.conf.urls import url
 
-from .views import EjabberdServiceView, JsxcView
+from .views import EjabberdServiceView
 
 
 urlpatterns = [
-    url(r'^apps/xmpp/$', EjabberdServiceView.as_view(), name='index'),
-    url(r'^apps/xmpp/jsxc/$', JsxcView.as_view(), name='jsxc'),
+    url(r'^apps/xmpp/ejabberd/$', EjabberdServiceView.as_view(), name='index')
```
Comment (jvalleroy) [inline plinth/modules/ejabberd/urls.py]: Drop 'xmpp' from url.

File: plinth/modules/jsxc/__init__.py [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+from django.utils.translation import ugettext_lazy as _
+import logging
+import socket
+
+from plinth import actions
+from plinth import action_utils
+from plinth import cfg
+from plinth import frontpage
+from plinth import service as service_module
+from plinth.signals import pre_hostname_change, post_hostname_change
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: No need to import these signals.

File: plinth/modules/jsxc/__init__.py [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+from plinth.signals import domainname_change
+
+
+version = 2
+
+depends = ['apps']
+
+managed_packages = ['libjs-jsxc']
+
+title = _('JSXC')
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: Change to 'Chat Client \n (JSXC)' to match other modules.

File: plinth/modules/jsxc/__init__.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+
+description = [
+
+    _('To actually communicate, you can use the web client or any other '
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: How about this:
"JSXC is a web client for XMPP. Typically it is used with an XMPP server running locally."

File: plinth/modules/jsxc/__init__.py [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+      '>XMPP client</a>.'),
+]
+
+service = None
+
+logger = logging.getLogger(__name__)
+
+
+def init():
+    """Initialize the XMPP module"""
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: XMPP -> JSXC

File: plinth/modules/jsxc/__init__.py [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+
+    global service
+    setup_helper = globals()['setup_helper']
+    if setup_helper.get_state() != 'needs-setup':
+        service = service = service_module.Service(
+            'jsxc', title, ports=['http', 'https'], is_external=True,
+            is_enabled=is_enabled, enable=enable, disable=disable)
+        if is_enabled():
+            add_shortcut()
+
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: This module needs a setup function. It should install the managed_packages, add the shortcut, and initialize service if needed.

File: plinth/modules/jsxc/__init__.py [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+def add_shortcut():
+    frontpage.add_shortcut('jsxc', _('Chat Client \n (jsxc)'),
+                           url=reverse_lazy('xmpp:jsxc'),
+                           login_required=True)
+
+
+def is_enabled():
+    """Return whether the module is enabled."""
+    return action_utils.webserver_is_enabled('jwchat-plinth'))
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: The jwchat-plinth apache conf is sort of misnamed. It is actually an interface to the ejabberd's BOSH server, so it is handled by ejabberd module.

I'm not sure if JSXC module really needs enable/disable. For now I would suggest, have the is_enabled simply check if JSXC module has completed setup (like lines 62-63 above).

File: plinth/modules/jsxc/__init__.py [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+
+def get_domainname():
+    """Return the domainname"""
+    fqdn = socket.getfqdn()
+    return '.'.join(fqdn.split('.')[1:])
+
+
+def enable():
+    """Enable the module."""
+    actions.superuser_run('jsxc', ['enable'])
```
Comment (jvalleroy) [inline plinth/modules/jsxc/__init__.py]: There's no action script named 'jsxc'. Actually I don't think this is needed, so we can remove this line and line 96.

File: plinth/modules/jsxc/templates/jsxc.html [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+
+{% block description %}
+
+  {% for paragraph in description %}
+    <p>{{ paragraph|safe }}</p>
+  {% endfor %}
+
+
+  <p>
+    <a href='{% url "xmpp:jsxc" %}' target='_blank' class='btn btn-primary'>
```
Comment (jvalleroy) [inline plinth/modules/jsxc/templates/jsxc.html]: URL should be referring to jsxc module, e.g. "jsxc:jsxc".

File: plinth/modules/jsxc/urls.py [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+URLs for the JSXC module
+"""
+
+from django.conf.urls import url
+
+from .views import JSXCServiceView, JsxcView
+
+
+urlpatterns = [
+    url(r'^apps/jsxc/$', JsxcView.as_view(), name='jsxc'),
```
Comment (jvalleroy) [inline plinth/modules/jsxc/urls.py]: Need 2 URLs:
- apps/jsxc for JSXCServiceView ('index')
- apps/jsxc/jsxc for JsxcView ('jsxc')

File: plinth/modules/jsxc/views.py [hunk 28]
```diff
@@ -16,27 +16,27 @@
...
 
 
-class EjabberdServiceView(ServiceView):
+class JSXCServiceView(ServiceView):
     """Show ejabberd as a service."""
-    service_id = xmpp.managed_services[0]
-    template_name = 'xmpp.html'
-    description = xmpp.description
-    diagnostics_module_name = 'xmpp'
+    service_id = jsxc.managed_services[0]
```
Comment (jvalleroy) [inline plinth/modules/jsxc/views.py]: jsxc has not managed_services. I would just use 'jsxc' string constant here.

File: plinth/modules/jsxc/views.py [hunk 29]
```diff
@@ -16,27 +16,27 @@
...
+    template_name = 'jsxc.html'
+    description = jsxc.description
+    diagnostics_module_name = 'jsxc'
```
Comment (jvalleroy) [inline plinth/modules/jsxc/views.py]: I think no diagnostics are needed, so I would remove this line.

**Comment by 500swapnil** [comment]
@jvalleroy I have made the changes you suggested. Please have a look!

**Comment by jvalleroy** [comment]
@500swapnil Thanks, I've merged this change with some fixes.

**Comment by 500swapnil** [comment]
Thanks! I'll work on other issues now!

**Pull request closed without merging**